### ISSUE TITLE
DLND-Week5

### ISSUE THREAD

**Issue by chenghong-lin-nu** [Hclmaster/blog#13]
## Generative Adversarial Networks(生成对抗网络)
- They are ML model that can image new things.

### Application(应用)
- GAN可以用于生成逼真的数据
- 例如说STACKGAN Model，它可以接收文本形式的描述，然后根据那个描述来生成相应的图片（这些图片都是被新创造出来的，不是从数据库中进行寻找 ）。
- 还有Pix2Pix，这个是为了艺术家而设计的，当艺术家画了一个比较抽象的图像时，然后这个模型就会寻找出与之最为相似的图片。

### How GANs work
- GANs模型能够让我们一次性生成整幅图像。
- Generative Network（生成网络） 将随机噪声作为输入。
- 然后将此噪声传入可微函数，转变和重塑它，使之具有可识别的结构。
- 该生成网络的输出是一个逼真的图像。
- 随机输入噪音的选择决定了哪种图像会从generator network中输出。
- 对于Generative Model来说，它是unsupervised learning（非监督式学习）。
![image](https://user-images.githubusercontent.com/16880879/39092219-b394982c-463b-11e8-8e84-62b80b4869ca.png)
- 我们只是向模型展示大量图像，然后要求网络生成更多符合相同概率分布的图像。
![image](https://user-images.githubusercontent.com/16880879/39092240-213743fc-463c-11e8-9ac3-97a7d1eedd4c.png)
- GANs使用辨别器（discriminator）学会引导生成器。它只是一个普通的神经网络分类器。
- 然后喂它一半来自训练数据集的真实数据，一半来自于生成器的假的图像。
- discriminator试图去输出一个输入图片是否为真的概率值。对于真实的图像，它试图分配一个接近1的概率；对于假图像，分配一个接近0的概率。
- 然后generator却试图去生成更多的能够让辨别器认为是1（也就是真实的）图片，相当于是**生成假钞一样**，总要越做越像。
- 这样的话，generator就会被迫生成更多更逼真的图像来以假乱真。

### Games and Equilibria
- Adversarial在GANs中的意思就是指有两个network are in competition with each other，相互竞争。
- 我们依赖game theory来更好的理解这个理论。
- Equilibrium: 平等
- 在博弈论中，equilibrium是一个非常重要的情景，每个人都无法通过改变它们的策略来改变结果payoff。（这种情况在GANs中发生时是在：that happens when we are at the maximum of the discriminator and minimum of the generator，然后呢，that point is called the saddle point!）
![image](https://user-images.githubusercontent.com/16880879/39661189-41db46a0-5080-11e8-9780-7bc43f1552fe.png)
- 问题：但是下面那张图说 the game plotted here has an equilibrium in the middle，为什么中间的是equilibrium呢？
![image](https://user-images.githubusercontent.com/16880879/39661220-8842de32-5080-11e8-917a-ab93fd127f37.png)
- 后来它又说，the equilibrium is like a minimum of the value function.
- GANs有2个player，generataor和discriminator，each player has its own cost.
- 简单的来说，the cost of the discriminator is just the negative of the cost of the generator.
- GAN的核心问题就是在博弈中寻找到一个平衡。
- 传统的训练GANs的方法就是：train GANs by running 2 optimization algorithms simultaneously, each minimize one player's cost with respect to that player's parameters.
- 问题：为什么说GANs就是想找到一个在generator和discriminator之间的平衡呢？

### Practical tips and tricks for training GANs
- generator和discriminator里面都至少有一个hidden layer。
- 对于hidden units来说，大多数activation function都是可用的，但是leaky ReLU算是比较有用的。（它可以保证gradient can flow through the entire architecture）
- **the only way the generator can learn is to receive a gradient from the discriminator.**
![image](https://user-images.githubusercontent.com/16880879/39123573-ce33479c-472a-11e8-9f32-69da581fec2f.png)
- **a popular way for the output of the generator is the hyperbolic tangent activation function.所以对于它的输出而言，范围肯定在-1到1之间了。**
- **对于discriminator的output来说，输出需要是一个probability，所以这里我们使用的是sigmoid activation function。**
- GANs和其他的模型算法不同，因为它需要run two algorithms simultaneously.
- generator和discriminator的loss都是用AdamOptimizer的，对于真实的数据，最后输出为1；而假数据则接近于0。然后损失函数是sigmoid_cross_entropy。
![image](https://user-images.githubusercontent.com/16880879/39124229-3f51508e-472d-11e8-90bb-bf25be8ce136.png)
- 人们经常犯的一个错误是：they forget the numerically stable version of cross-entropy.
![image](https://user-images.githubusercontent.com/16880879/39124401-e683ede4-472d-11e8-82ec-f3f7c4bdeafe.png)
- 这里的logits就是discriminator产生的输出（outputs）。它是在放入sigmoid之前被产生的。
- 然后对于**Generator的Lost**，cross_entropy_loss又有些不同。
- 它这里labels是进行翻转过的；比如说0变成1，然后1变成0.
- the generator will **maximize the log probability of the wrong labels**.
![image](https://user-images.githubusercontent.com/16880879/39663696-936a039c-50aa-11e8-9a73-529154cf49c8.png)
- 很多人都会把负的discriminator loss as the expression for the generator loss.
- 对于一般的CNN来说（也就是当我们构建Classifier的CNN时），都是一开始输入比较大的图形，然后经过卷积的处理之后就变成了一个窄又长的模型。
- 但是当CNN作为一个generator net的时候，我们做的就是相反的事情了。
- 一开始是short images，然后变成高大的图像。
![image](https://user-images.githubusercontent.com/16880879/39663787-2400c264-50ac-11e8-9b1f-d446d6461ce2.png)
- 后面的一小部分还是需要补一下。。。
